Add Guide to retrieve controllers addresses + their permissions (#313)
lukso-network/docs: docs/guides/key-manager/get-controllers.md

---
sidebar_label: 'Get controller addresses'
sidebar_position: 2
---

# Get controller addresses

In this guide, you will learn how to:

- retrieve all the addresses that have some permissions set on a Universal Profile,
- retrieve the permission of each of these **controller addresses**.

![Get controller addresses](/img/standards/lsp6/lsp6-address-permissions-array.jpeg)

We will use the [_erc725.js_](../../tools/erc725js/getting-started.md) library to do this with just over 30 lines of code.

## Setup

```bash
npm install erc725.js
```

## Step 1 - setup web3.js and erc725.js

> The first step is to setup both and _web3.js_ and _erc725.js_, connected to the LUKSO L14 network. We will also need the address of the Universal Profile that we want to get the controller addresses from.

Reviewer: ```suggestion
The first step is to set up both _web3.js_ and _erc725.js_, and connect to the LUKSO L14 network. We will also need the address of the Universal Profile that we want to get the controller addresses from.
```

```js
import { ERC725 } from '@erc725/erc725.js';
import LSP6Schema from '@erc725/erc725.js/schemas/LSP6KeyManager.json';
import Web3 from 'web3';

// setup
const myUniversalProfileAddress = '0xC26508178c4D7d3Ad43Dcb9F9bb1fab9ceeD58B5';
const RPC_ENDPOINT = 'https://rpc.l16.lukso.network';

// step 1 - setup erc725.js
const web3 = new Web3(RPC_ENDPOINT);
const erc725 = new ERC725(
  LSP6Schema,
  myUniversalProfileAddress,
  web3.currentProvider,
);
```

## Step 2 - Get the list of controller addresses

> The next step is to retrieve the list of controller addresses. In other words, addresses that have some permissions set and are allowed to interact with the Universal Profile.

Reviewer: ```suggestion
The next step is to retrieve the list of controller addresses. In other words, addresses with some permissions can interact with the Universal Profile.
```

> All the controller addresses are listed under the data key `AddressPermissions[]`. We can therefore obtain this list of addresses by:

Reviewer: ```suggestion
All the controller addresses are listed under the data key `AddressPermissions[]`. The list of addresses can be obtained in two steps:
```

Reviewer: I would suggest keeping bulletpoints/orders in present.

Author: I think having _"listed"_ and then the word _"Iist of"_ makes it a bit repetitive in the end.
My suggestion:
```suggestion
All the controller addresses are listed under the data key `AddressPermissions[]`. We can retrieve them by:
```

Reviewer: Good point, i like it.

> 1. querying the `AddressPermissions[]` to get the array length, and know **how many controller addresses** have some permissions on the UP.

Reviewer: ```suggestion
1. query the `AddressPermissions[]` to get the array length and know **how many controller addresses** have some permissions on the UP.
```

Reviewer 2: I'd like to keep `querying` and remove the comma.

> 2. querying each index of the `AddressPermissions[index]` array to retrieve each address individually.

Reviewer: ```suggestion
2. query each index of the `AddressPermissions[index]` array to retrieve each address individually.
```

Reviewer 2: I think `querying` is better than `query`

"this list of addresses by query the" doesn't sounds correct.

Reviewer: true, would either reorg the sentence structure to have it in present, or leave present progressive then. whatever is more fluent

> Thanks to _erc725.js_, we can do that very easily. The library will do both steps for us and will return the full list of controller addresses.

Reviewer: 2x will in a row

```suggestion
Thanks to _erc725.js_, we can do that very easily. The library will do both steps for us and returns the full list of controller addresses.
```

```js
// step 2 - get the list of addresses that have permissions on the Universal Profile
const result = await erc725.getData('AddressPermissions[]');
console.log(result);

// {
//   key: '0xdf30dba06db6a30e65354d9a64c609861f089545ca58c6b4dbe31a5f338cb0e3',
//   name: 'AddressPermissions[]',
//   value: [
//     '0x5F606b5b237623463a90F63230F9b929321dbCBa',
//     '0xa1061408e55c971fD129eF5561dFB953d598dAD7'
//   ]
// }
```

## Step 3 - Get permissions of each controllers

> Getting the addresses of each controller is not enough. We need to retrieve the permissions of each of these controller addresses, so to know what they are allowed to do in our Universal Profile.

Reviewer: ```suggestion
Getting the addresses of each controller is not enough. We need to retrieve the permissions of each controller address to know what they are allowed to do in our Universal Profile.
```
